File: app-node/src/graph.ts
import { StateGraph, END } from "@langchain/langgraph";
import { BaseMessage, SystemMessage, ToolMessage, AIMessage } from "@langchain/core/messages";
import { ChatOpenAI } from "@langchain/openai";
import { StructuredTool } from "@langchain/core/tools";

// State definition
interface AgentState {
    messages: BaseMessage[];
}

export function buildGraph(llm: ChatOpenAI, tools: StructuredTool[]) {
    // LLM with tools
    const llmWithTools = llm.bindTools(tools);
    const toolsMap = Object.fromEntries(tools.map(t => [t.name, t]));

    // Node definitions
    async function callLlm(state: AgentState) {
        console.log("\n--- [Agent: Thinking] ---");
        const messages = state.messages;
        const systemMessage = new SystemMessage("Answer questions using tools if needed. If using a PDF tool, cite content. Be concise.");

        const response = await llmWithTools.invoke([systemMessage, ...messages]);

        if (response.content) {
            console.log(`[Agent: Response] ${response.content.toString().substring(0, 100)}${response.content.toString().length > 100 ? '...' : ''}`);
        }

        return { messages: [response] };
    }

    async function takeAction(state: AgentState) {
        const lastMessage = state.messages[state.messages.length - 1] as AIMessage;
        if (!lastMessage.tool_calls || lastMessage.tool_calls.length === 0) {
            return { messages: [] };
        }

        const results = [];
        console.log(`--- [Agent: Taking Action] Using ${lastMessage.tool_calls.length} tool(s) ---`);

        for (const call of lastMessage.tool_calls) {
            const tool = toolsMap[call.name];
            if (tool) {
                console.log(`[Tool Call: ${call.name}] Args: ${JSON.stringify(call.args)}`);
                const output = await tool.invoke(call.args);
                const content = typeof output === 'string' ? output : JSON.stringify(output);

                console.log(`[Tool Result: ${call.name}] Output length: ${content.length} chars`);

                results.push(new ToolMessage({
                    tool_call_id: call.id!,
                    name: call.name,
                    content: content
                }));
            } else {
                console.warn(`[Agent: Warning] Tool ${call.name} not found!`);
            }
        }
        return { messages: results };
    }

    function shouldContinue(state: AgentState) {
        const lastMessage = state.messages[state.messages.length - 1] as AIMessage;
        if (lastMessage.tool_calls && lastMessage.tool_calls.length > 0) {
            return "tools";
        }
        return END;
    }

    const workflow = new StateGraph<AgentState>({
        channels: {
            messages: {
                reducer: (x: BaseMessage[], y: BaseMessage[]) => x.concat(y),
                default: () => [],
            }
        }
    })
        .addNode("llm", callLlm)
        .addNode("tools", takeAction)
        .addEdge("tools", "llm")
        .addConditionalEdges("llm", shouldContinue)
        .setEntryPoint("llm");

    return workflow.compile();
}

File: app/graph.py
from typing import TypedDict, Sequence, Annotated
from operator import add as add_messages
from langgraph.graph import StateGraph, END
from langchain_core.messages import BaseMessage, SystemMessage, ToolMessage
from langchain_openai import ChatOpenAI

class AgentState(TypedDict):
    messages: Annotated[Sequence[BaseMessage], add_messages]


def build_graph(llm, tools):

    tools_dict = {t.name: t for t in tools}

    def should_continue(state: AgentState):
        last = state["messages"][-1]
        return hasattr(last, "tool_calls") and len(last.tool_calls) > 0

    def call_llm(state: AgentState):
        system = SystemMessage(
            content="Answer questions using the PDF. Cite content."
        )
        response = llm.invoke([system] + list(state["messages"]))
        return {"messages": [response]}

    def take_action(state: AgentState):
        tool_calls = state["messages"][-1].tool_calls
        results = []

        for call in tool_calls:
            tool = tools_dict[call["name"]]
            result = tool.invoke(call["args"]["query"])
            results.append(
                ToolMessage(
                    tool_call_id=call["id"],
                    name=call["name"],
                    content=result
                )
            )

        return {"messages": results}

    graph = StateGraph(AgentState)
    graph.add_node("llm", call_llm)
    graph.add_node("tools", take_action)

    graph.add_conditional_edges(
        "llm",
        should_continue,
        {True: "tools", False: END}
    )

    graph.add_edge("tools", "llm")
    graph.set_entry_point("llm")

    return graph.compile()